Add OL 21.0.0.3 GA Release Blog to Staging (#1216)
OpenLiberty/blogs: posts/2021-03-19-microprofile-4-21003.adoc

---
layout: post

> title: "A new MicroProfile milestone now available in Open Liberty 21.0.0.3"

Reviewer: Can we change this to be either "MicroProfile 4.0 now available in Open Liberty 21.0.0.3" or "Full MicroProfile 4.0 support now available in Open Liberty 21.0.0.3"?

Author: Updated to "Full MicroProfile 4.0 support now available in Open Liberty 21.0.0.3".

categories: blog
author_picture: https://avatars3.githubusercontent.com/austin0
author_github: https://github.com/austin0

> seo-title: A new MicroProfile milestone now available in Open Liberty 21.0.0.3 - OpenLiberty.io

Reviewer: Same as comment for title above (please make the same)

Author: Updated

> seo-description: With updates to all 8 MicroProfile components, MicroProfile 4.0 is now generally available in Open Liberty 21.0.0.3.

Reviewer: I really liked the previous title, but could we add it here instead? So perhaps "A new MicroProfile milestone is now available in Open Liberty. With updates to all 8 MicroProfile components (Config, Fault Tolerance, Health, JWT RBAC, Metrics, Open API, Open Tracing and Rest Client), full support for MicroProfile 4.0 is
 available in Open Liberty 21.0.0.3."
I've listed out the different APIs here so that if anyone searches for them hopefully our article will also appear. Let me know if you disagree with including these though.

Author: Sounds really good, thanks.

> blog_description: With updates to all 8 MicroProfile components, MicroProfile 4.0 is now generally available in Open Liberty 21.0.0.3.

Reviewer: Please make this the same as seo-description

Author: Updated.

open-graph-image: https://openliberty.io/img/twitter_card.jpg
---

> = A new MicroProfile milestone now available in Open Liberty 21.0.0.3

Reviewer: Please change this to new title

Author: Updated.

Austin Bailey <https://github.com/austin0>
:imagesdir: /
:url-prefix:
:url-about: /

// tag::intro[]

> In 21.0.0.3, we have an milestone update to the MicroProfile feature with a major version increase to 4.0. This includes updates to all 8 MicroProfile components. To get it, run your app with the latest version of Open Liberty.

Reviewer: "In 21.0.0.3, we have a milestone update to the MicroProfile feature with a major version increase to 4.0. This includes updates to all 8 MicroProfile components: Config, Fault Tolerance, Health, JWT RBAC, Metrics, Open API, Open Tracing and Rest Client. To utilize these updated components, run your app with the latest version of Open Liberty."
I have changed an to a for grammatical reasons. I have also added in the different component names, as per my comment above.

Author: Added, thanks.

> If you're feeling curious about what's coming soon in Open Liberty, take a look at the 'beta status' features in the latest development builds, for example LDAP connection support for Kerberos authentication is now available in the Open Liberty 21.0.0.4-beta release. 

Reviewer 2: are we posting the beta blog at the same time? can we add a link to that?

Reviewer: agree, could we add a link to that here please?

Reviewer: I feel like this short paragraph about the beta should come after the part below listing what's new in this GA release.

Author: Link added.

In link:{url-about}[Open Liberty] 21.0.0.3:

* <<MP4, MicroProfile 4.0>>

View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3ARELEASE_VERSION+label%3A%22release+bug%22[21.0.0.3].
// end::intro[]

// tag::run[]
[#run]

== Run your apps using 21.0.0.3

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>21.0.0.3</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[21.0.0.3,)'
}
----

Or if you're using Docker:

[source]
----
FROM open-liberty
----
//end::run[]

Or take a look at our link:{url-prefix}/downloads/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

//tag::features[]

[#MP4]

> === MicroProfile 4.0

Reviewer: Could we add support on the end here please?
So "MicroProfile 4.0 Support"?

Author: Added.

> Microprofile 4.0 is a major version update complete with updates to the following 8 MicroProfile components:

Reviewer: MicroProfile 4.0 is the first release under the Working Group model and MicroProfile Specification Process and alligns with Jakarta EE 8 (add link here). This major version update includes updates to the following 8 MicroProfile components:

Author: Updated.

* MicroProfile Config 2.0

* MicroProfile Fault Tolerance 3.0

* MicroProfile Health 3.0

* MicroProfile JWT Authentication 1.2

* MicroProfile Metrics 3.0

* MicroProfile OpenAPI 2.0

* MicroProfile OpenTracing 2.0

* MicroProfile Rest Client 2.0

> As this is such a large update standalone blogs have been written for both MicroProfile 4.0 and a few select components, a deep-dive into MicroProfile 4.0 can be found in the blog linked below with further component-specific blogs due to release over the coming weeks:

Reviewer 2: ```suggestion
As this is such a large update, separate blog posts have been written for both MicroProfile 4.0 and a few of the individual components. A deep-dive into MicroProfile 4.0 can be found in the post linked below with further component-specific posts due to release over the coming weeks:
```

Author: Changed.

> As this is such a large update standalone blogs have been written for both MicroProfile 4.0 and a few select components, a deep-dive into MicroProfile 4.0 can be found in the blog linked below with further component-specific blogs due to release over the coming weeks:

Reviewer: please add a comma after update. So "update, standalone blogs...".
Please switch the comma after components for a full-stop and create new sentence. So, "components. A deep-dive...."
"due to release" could we change this to "being released over the coming weeks"

Author: Pretty much exactly what Tom commented :)

> * link:https://openliberty.io/blog/2021/03/19/microprofile40-open-liberty-21003.html[MicroProfile 4.0 support in Open Liberty 21.0.0.3]

Reviewer: Could we make this text slightly bigger for this link so it's easier to see?

Author: I've taken the bullet point away so it should appear larger, again can see how it looks when build and try to go bigger.

> 

Reviewer: could we have a subtitle here to separate it from the section above? Perhaps just something simple like "Making use of MicroProfile 4.0"?

Author: Added, can see if we like if after it builds again 👍

To make use of MicroProfile 4.0 you must add the feature to both `pom.xml` and `server.xml`:

Add the following to your `server.xml`:
[source, xml]
----
<server>
  <featureManager>
    <feature>microProfile-4.0</feature>
  </featureManager>
</server>
----

And add the following to your `pom.xml`:
[source, xml]
----
<dependency>
    <groupId>org.eclipse.microprofile</groupId>
    <artifactId>microprofile</artifactId>
    <version>4.0.1</version>
    <type>pom</type>
    <scope>provided</scope>
</dependency>
----

//end::features[]

[#bugs]
== Significant bugs fixed in this release

> We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A21003+label%3A%22release+bug%22[full list of bugs fixed in 21.0.0.3].

Reviewer: As well as adding support for MicroProfile 4.0, we've also spent some time fixing bugs.
Could we add this at the beginning here just to link the two together a little more?

Author: Added.

* link:https://github.com/OpenLiberty/open-liberty/issues/15927[Fix the injection of Optional Lists and Sets (MP Config)]
+

> Previously in Open Liberty it was not possible to inject an `Optional<List<String>>` or `Optional<Set<String>>` using `@ConfigProperty`. This was problematic as in the newly release MP Config 2.0 removed the ability to inject an empty list as an empty value, as it will treated the same as a missing value. This would lead to diffuclties when writing backwards-compatible code for both MP Config 2.0 and 1.x versions. This issue has now been fixed and there should be no issues injecting optional lists, for more information view the link:https://openliberty.io/docs/latest/reference/feature/mpConfig-2.0.html[MP Config 2.0 documentation].

Reviewer 2: ```suggestion
Previously, a bug in Open Liberty's mpConfig-1.x features meant it was not possible to inject an `Optional<List<String>>` or `Optional<Set<String>>` using `@ConfigProperty`. This issue has now been fixed and there should be no issues injecting optional Lists and Sets.
```

Author: Updated, thanks.

> Previously in Open Liberty it was not possible to inject an `Optional<List<String>>` or `Optional<Set<String>>` using `@ConfigProperty`. This was problematic as in the newly release MP Config 2.0 removed the ability to inject an empty list as an empty value, as it will treated the same as a missing value. This would lead to diffuclties when writing backwards-compatible code for both MP Config 2.0 and 1.x versions. This issue has now been fixed and there should be no issues injecting optional lists, for more information view the link:https://openliberty.io/docs/latest/reference/feature/mpConfig-2.0.html[MP Config 2.0 documentation].

Reviewer: this was problematic, as the ..... (please add comma here and delete in).
"newly released" - please make this past tense
"as it will treated the same as a missing value." - this sounds odd to me. Could we perhaps change this to "as it is treated the same as a missing value."?
could we switch fixed for resolved?
"and so there should now be no issues injecting...." (I've added so and now here)

Author: Section has been overhauled by Tom.

* link:https://github.com/OpenLiberty/open-liberty/issues/12215[Read jvm.options file from the intended directory]
+

> According to the link:[Open Liberty documentation] the `jvm.options` file is inteded to be read from the `${wlp.user.dir}/shared/jvm.options` directory to allow for custom JVM options to be loaded. Prior to Open Liberty 21.0.0.3 this file was erroneously fetched from the `${WLP_INSTALL_DIR}/usr/shared/jvm.options` path, this behaviour has now been corrected.

Reviewer 2: ```suggestion
According to the link:[Open Liberty documentation] the `jvm.options` file is inteded to be read from the `${wlp.user.dir}/shared/jvm.options` directory to allow for custom JVM options to be loaded. Prior to Open Liberty 21.0.0.3, this file was erroneously fetched from the `${WLP_INSTALL_DIR}/usr/shared/jvm.options` path. This behaviour has now been corrected.
```

Author: Updated.

> According to the link:[Open Liberty documentation] the `jvm.options` file is inteded to be read from the `${wlp.user.dir}/shared/jvm.options` directory to allow for custom JVM options to be loaded. Prior to Open Liberty 21.0.0.3 this file was erroneously fetched from the `${WLP_INSTALL_DIR}/usr/shared/jvm.options` path, this behaviour has now been corrected.

Reviewer: inteded should be intended (spelling mistake)

Author: Thanks for catching.

* link:https://github.com/OpenLiberty/open-liberty/issues/15853[Update Netty to 4.1.59.Final (gRPC Client 1.0)]
+

> Due to a vulnerability discovered in Netty client/server framework a security update has been pushed to Open Liberty 21.0.0.3, going from version 4.1.52.Final to 4.1.59.Final. The discovered vulnerability (link:https://github.com/advisories/GHSA-5mcr-gq6c-3hq2[CVE-2021-21290]) would not have affected any Open Liberty code but the precautionary measure to update has been taken to maintain high security standards.

Reviewer: Should Netty have a link attached?

Author: Link added.

> * link:https://github.com/OpenLiberty/open-liberty/issues/15744[Enable MyFaces 2.3.8 for Open Liberty]

Reviewer 2: ```suggestion
* link:https://github.com/OpenLiberty/open-liberty/issues/15744[Update to Apache MyFaces 2.3.8]
```

Author: Updated.

+
The `jsf-2.3` feature in Open Liberty makes use of link:https://myfaces.apache.org/core23/index.html[Apache MyFaces] JavaServer Faces implementation. With the release of Apache MyFaces 2.3.8 a number of improvements and bug fixes have been made, for more information visit the link:https://issues.apache.org/jira/secure/ReleaseNote.jspa?projectId=10600&version=12349324[Apache MyFaces 2.3.8 changelog].

* link:https://github.com/OpenLiberty/open-liberty/issues/11249[Fix memory leak when applications do not close JAXRS Client objects]
+

> Prior to Open Liberty 21.0.0.3 an issue within JAXRS existed where `Client` objects would not be properly disopsed of when their use had expired, thus causing a memory leak. The cause of this was `JAXRSClientImpl` keeping a hard reference to the client so preventing it from being picked up by garbage collection; this behaviour has since been corrected. For more information view the link:https://openliberty.io/docs/20.0.0.12/reference/feature/jaxrsClient-2.1.html[JAX-RS 2.1 documentation]. 

Reviewer 2: ```suggestion
Prior to Open Liberty 21.0.0.3, an issue within JAXRS existed where `Client` objects would not be properly disopsed of when their use had expired, thus causing a memory leak. The cause of this was `JAXRSClientImpl` keeping a hard reference to the client so preventing it from being picked up by garbage collection; this behaviour has since been corrected. For more information view the link:https://openliberty.io/docs/20.0.0.12/reference/feature/jaxrsClient-2.1.html[JAX-RS 2.1 documentation]. 
```

Author: Added, thanks.
